File: AkkaUnitTest/src/main/java/org/akka/essentials/unittest/actors/SupervisorActor.java
package org.akka.essentials.unittest.actors;

import static akka.actor.SupervisorStrategy.escalate;
import static akka.actor.SupervisorStrategy.resume;
import static akka.actor.SupervisorStrategy.stop;
import scala.concurrent.duration.Duration;
import akka.actor.ActorRef;
import akka.actor.OneForOneStrategy;
import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.actor.SupervisorStrategy.Directive;
import akka.actor.UntypedActor;
import akka.japi.Function;


public class SupervisorActor extends UntypedActor {

	private ActorRef childActor;

	public ActorRef getChildActor() {
		return childActor;
	}

	private static SupervisorStrategy strategy = new OneForOneStrategy(10,
			Duration.create("10 second"), new Function<Throwable, Directive>() {
				public Directive apply(Throwable t) {
					if (t instanceof IllegalArgumentException) {
						return stop();
					} else if (t instanceof NullPointerException) {
						return resume();
					} else
						return escalate();
				}
			});

	@Override
	public SupervisorStrategy supervisorStrategy() {
		return strategy;
	}

	public void onReceive(Object o) throws Exception {
		if (o instanceof Props) {
			this.childActor = getContext().actorOf((Props) o, "childActor");
			sender().tell(childActor);
		} else
			childActor.tell(o, sender());
	}
}

File: AkkaUnitTest/src/main/scala/org/akka/essentials/unittest/example/SupervisorActor.scala
package org.akka.essentials.unittest.example
import akka.actor.SupervisorStrategy.Escalate
import akka.actor.SupervisorStrategy.Resume
import akka.actor.SupervisorStrategy.Stop
import akka.actor.actorRef2Scala
import akka.actor.ActorRef
import akka.actor.Actor
import akka.actor.OneForOneStrategy
import akka.actor.Props
import akka.util.duration.intToDurationInt

class SupervisorActor() extends Actor {

  var childActor: ActorRef = _

  def getChildActor() = childActor

  def receive: Receive = {
    case message: Props =>
      childActor = context.actorOf(message, name = "childActor")
      sender ! childActor
    case message => 
      childActor.tell(message, sender)
  }

  override val supervisorStrategy = OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
    case _: NullPointerException => Resume
    case _: IllegalArgumentException => Stop
    case _: Exception => Escalate
  }
}